update_approver/main.py

```diff
@@ -0,0 +1,187 @@
+approve_updates = config('APPROVER_APPROVE_UPDATES', default='false', cast=bool)
+
+
+class SubscriptionWatcher(threading.Thread):
+    '''Watch for events on Subscription resources'''
+
+    def __init__(self, q, client):
+        self.q = q
+        self.client = client
+        super().__init__(daemon=True)
+
+    def run(self):
+        for event in self.client.subscriptions.watch():
+            self.q.put(('subscription', event))
+
+
+class FileWatcher(threading.Thread):
+    '''Watch for filesystem events in config_dir'''
+
+    def __init__(self, q):
+        self.q = q
+        super().__init__(daemon=True)
+
+    def run(self):
+        for event in watch(config_dir):
+            self.q.put(('fs', event))
+
+
+class Client(DynamicClient):
+    def __init__(self):
+        try:
+            kubernetes.config.load_incluster_config()
+        except kube_exc.ConfigException:
+            kubernetes.config.load_kube_config()
+
+        k8s = kubernetes.client.ApiClient()
+        super().__init__(k8s)
+
+    @cached_property
+    def subscriptions(self):
+        return self.resources.get(
+            api_version='operators.coreos.com/v1alpha1',
+            kind='Subscription'
+        )
+
+    @cached_property
+    def installplans(self):
+        return self.resources.get(
+            api_version='operators.coreos.com/v1alpha1',
+            kind='InstallPlan'
+        )
+
+    def process_update_spec(self, specfile: Path):
+        with specfile.open('r') as fd:
+            spec = yaml.safe_load(fd)
+
+        LOG.info('processing subscription {} in namespace {}',
+                 spec['name'],
+                 spec['namespace'])
+
+        # look up named subscription
+        sub = self.subscriptions.get(name=spec['name'], namespace=spec['namespace'])
+
+        # get current installplan
+        plan_name = sub.status.installPlanRef.name
+        LOG.debug('{}: got installplan {}',
+                  spec['name'],
+                  plan_name)
+        plan = self.installplans.get(name=plan_name, namespace=spec['namespace'])
+        have = plan.spec.clusterServiceVersionNames[0]
+
+        # check if installplan matches request version
+        if have != spec['version']:
+            LOG.warning('{}: invalid version: have={}, want={}',
+                        spec['name'],
+                        have,
+                        spec['version'])
+            return
+
+        # check if installplan was previously approved
+        if plan.spec.approved:
+            LOG.info('{}: version {} is already approved',
+                     spec['name'],
+                     spec['version'])
+            return
+
+        # approve the plan
+        LOG.warning('{}: approve version {}{}',
+                    spec['name'],
+                    spec['version'],
+                    '' if approve_updates else ' (dry run)')
+
+        if approve_updates:
+            plan.spec.approved = True
+            self.intallplans.patch(
```

Comment:
shouldn't this say `self.installplans.get` ?

Reply:
No, we're patching the resource (setting `approved` to `true`).

Reply:
I was referring more to the typo `self.intallplans.patch` :)

Reply:
Ah, that makes more sense! :) Yeah, this is the one line of code that hasn't been run live yet and it shows...